src/client/producer.rs

```diff
@@ -374,130 +373,144 @@ where
+        tokio::select! {
+            _ = tokio::time::sleep(self.linger) => {},
+            r = result_slot.receive() => return extract(r, tag),
         }
 
         // Linger expired - reacquire lock
-        let mut inner = self.inner.lock().await;
+        let inner = self.lock().await;
 
         // Whilst holding lock - check hasn't been flushed already
         //
         // This covers two scenarios:
         // - the linger expired "simultaneously" with the publish
         // - the linger expired but another thread triggered the flush
         if let Some(r) = result_slot.peek() {
+            debug!(client=?self.client, ?tag, "Already flushed");
             return extract(r, tag);
         }
 
-        debug!("Linger expired - flushing");
+        debug!(client=?self.client, ?tag, "Linger expired - flushing");
 
         // Flush data
-        Self::flush_impl(&mut inner, self.client.as_ref(), self.compression).await;
+        Self::flush_impl(inner, Arc::clone(&self.client), self.compression).await?;
 
-        // We need to poll the result slot here to make the result available via `peek`, otherwise the next thread in
-        // this critical section will flush the producer a 2nd time. We are using an ordinary `.await` call here instead
-        // of `now_or_never` because tokio might preempt us (or to be precise: might preempt polling from the result slot).
-        extract(&result_slot.await, tag)
+        extract(result_slot.peek().unwrap(), tag)
     }
 
     /// Flushed out data from aggregator.
-    pub async fn flush(&self) {
-        let mut inner = self.inner.lock().await;
-        debug!("Manual flush");
-        Self::flush_impl(&mut inner, self.client.as_ref(), self.compression).await;
+    pub async fn flush(&self) -> Result<()> {
+        let inner = self.lock().await;
+        debug!(client=?self.client, "Manual flush");
+        Self::flush_impl(inner, Arc::clone(&self.client), self.compression).await?;
+        Ok(())
     }
 
     /// Flushes out the data from the aggregator, publishes the result to the result slot,
     /// and creates a fresh result slot for future writes to use
     async fn flush_impl(
-        inner: &mut ProducerInner<A>,
-        client: &dyn ProducerClient,
+        mut inner: OwnedMutexGuard<ProducerInner<A>>,
+        client: Arc<dyn ProducerClient>,
         compression: Compression,
-    ) {
-        trace!("Flushing batch producer");
+    ) -> Result<OwnedMutexGuard<ProducerInner<A>>> {
+        debug!(?client, "Flushing batch producer");
+
+        // Spawn a task to provide cancellation safety
+        let handle = tokio::spawn(async move {
+            // Swap the result slot first, this ensures that if the task panics the Drop
+            // implementation will still signal the result slot preventing flushing twice
+            let slot = std::mem::take(&mut inner.result_slot);
+
+            let (output, status_deagg) = match inner.aggregator.flush() {
```

Comment:
We still have an issue if the aggregator panics and leaves itself in an undefined state, but that's something for a future PR